model/DataBaseAccess.php

```diff
@@ -65,6 +67,88 @@ protected function getEventManager()
         return $this->getServiceLocator()->get(EventManager::SERVICE_ID);
     }
 
+
+    public function getPermissionsByUsersAndResources(array $userIds, array $resourceIds): array
+    {
+        /**
+         * @TODO FIXME Lets see if we can make this methods return better contract or if it is still necessary
+         */
+        // Permissions for an empty set of resources must be an empty array
+        if (empty($resourceIds)) {
+            return [];
+        }
+
+        $inQueryResources = implode(',', array_fill(0, count($resourceIds), '?'));
+        $inQueryUsers = implode(',', array_fill(0, count($userIds), '?'));
+
+        $query = <<<SQL
+SELECT
+    resource_id,
+    user_id,
+    privilege
+FROM data_privileges
+WHERE resource_id IN ($inQueryResources)
+  AND user_id IN ($inQueryUsers)
+SQL;
+
+        $results = $this->fetchQuery($query, [...$resourceIds, ...$userIds]);
+
+        // If resource doesn't have permission don't return null
+        $data = array_fill_keys($resourceIds, []);
+
+        foreach ($results as $result) {
+            $data[$result[self::COLUMN_RESOURCE_ID]][$result[self::COLUMN_USER_ID]][] = $result[self::COLUMN_PRIVILEGE];
+        }
+
+        return $data;
+    }
+
+    public function changeAccess(ChangeAccessCommand $command): void
+    {
+        $persistence = $this->getPersistence();
+        $class = $this;
+
+        $persistence->transactional(static function () use ($command, $persistence, $class): void {
+            foreach ($command->getUserIdsToRevokePermissions() as $userId) {
+                foreach ($command->getUserPermissionsToRevoke($userId) as $permission) {
+                    $resourceIds = $command->getResourceIdsByUserAndPermissionToRevoke($userId, $permission);
+
+                    if (count($resourceIds) > 0) {
+                        $persistence->exec(
+                            sprintf(
+                                'DELETE FROM data_privileges WHERE user_id = ? AND privilege = ? AND resource_id IN (%s)',
+                                implode(',', array_fill(0, count($resourceIds), ' ? ')),
```

Comment:
- [ ] We need to review if the query will work for 8k resources, the size of the query might be too big. We might have to split it on multiple queries.

Reply:
Done. Now it is deleted by chunks